### PR TITLE
Setup HDF5 dataset without having data ahead of time

### PR DESCRIPTION
**Pull request by ajtritt**
## Motivation

Add the ability to create a dataset on disk and get access to the handle to the dataset _without_ having to close and repoen a file. 

## How to test the behavior?
```python
dataio = H5DataIO(shape=(5,), dtype=int)
foo = Foo('foo1', dataio, "I am foo1", 17, 3.14)
bucket = FooBucket('bucket1', [foo])
foofile = FooFile(buckets=[bucket])

io = HDF5IO(self.path, manager=self.manager, mode='w')
io.write(foofile)

foo.my_data.dataset[:] = [0, 1, 2, 3, 4]
io.close()
```

## Checklist

- [x] Did you update CHANGELOG.md with your changes?
- [ ] Have you checked our [Contributing](https://github.com/hdmf-dev/hdmf/blob/dev/docs/CONTRIBUTING.rst) document?
- [ ] Have you ensured the PR clearly describes the problem and the solution?
- [x] Is your contribution compliant with our coding style? This can be checked running `flake8` from the source directory.
- [x] Have you checked to ensure that there aren't other open [Pull Requests](https://github.com/hdmf-dev/hdmf/pulls) for the same change?
- [ ] Have you included the relevant issue number using "Fix #XXX" notation where XXX is the issue number? By including "Fix #XXX" you allow GitHub to close issue #XXX when the PR is merged.

### CODE REVIEW

File: src/hdmf/backends/hdf5/h5_utils.py [hunk 1]
```diff
@@ -475,11 +475,20 @@ class H5DataIO(DataIO):
             {'name': 'allow_plugin_filters',
              'type': bool,
              'doc': 'Enable passing dynamically loaded filters as compression parameter',
-             'default': False}
+             'default': False},
+            {'name': 'shape',
+             'type': tuple,
+             'doc': 'the shape of the new dataset',
```
Comment (rly) [inline src/hdmf/backends/hdf5/h5_utils.py]:
```suggestion
             'doc': 'the shape of the new dataset, used only if data is None',
```

File: src/hdmf/backends/hdf5/h5_utils.py [hunk 2]
```diff
@@ -475,11 +475,20 @@ class H5DataIO(DataIO):
...
+             'default': None},
+            {'name': 'dtype',
+             'type': (str, type, np.dtype),
+             'doc': 'the data type of the new dataset',
```
Comment (rly) [inline src/hdmf/backends/hdf5/h5_utils.py]:
```suggestion
             'doc': 'the data type of the new dataset, used only if data is None',
```

File: src/hdmf/data_utils.py [hunk 3]
```diff
@@ -930,6 +934,16 @@ def data(self, val):
...
+    @property
+    def dtype(self):
+        """Get the wrapped data object"""
+        return self.__dtype
+
+    @property
+    def shape(self):
+        """Get the wrapped data object"""
+        return self.__shape
+
```
Comment (rly) [inline src/hdmf/data_utils.py]: This will break backward compatibility because currently `dtype` and `shape` are deferred to `self.data.dtype` and `self.data.shape`. I suggest one of these:
1. create a subclass of `DataIO` for empty datasets where `dtype` and `shape` are defined this way and use that
2. add an if statement in these: `if self.__dtype is not None, return self.__dtype, else, return self.__getattr__("dtype")`
3. change the name of these properties to something like `dtype_to_set` and `shape_to_set`

File: tests/unit/test_io_hdf5_h5tools.py [hunk 4]
```diff
@@ -1577,7 +1577,7 @@ def test_link(self):
 
             self.assertTrue(self.foo2.my_data.valid)  # test valid
             self.assertEqual(len(self.foo2.my_data), 5)  # test len
-            self.assertEqual(self.foo2.my_data.shape, (5,))  # test getattr with shape
+            self.assertEqual(self.foo2.my_data.size, (5,))  # test getattr with size
```
Comment (rly) [inline tests/unit/test_io_hdf5_h5tools.py]: I am surprised this works. When I put a breakpoint here, I can see that `self.foo2.my_data.size` is 5 ~~, and self.assertEqual(self.foo2.my_data.size, (5,)) raises an AssertionError~~. So why does this pass???

Comment (rly) [reply tests/unit/test_io_hdf5_h5tools.py]:
```
            self.assertEqual(self.foo2.my_data.size, (5,))  # test getattr with size
            self.assertEqual(self.foo2.my_data.size, 5)  # test getattr with size
>           self.assertEqual(5, (5,))  # test getattr with size
E           AssertionError: 5 != (5,)
```
That's weird.

Comment (rly) [reply tests/unit/test_io_hdf5_h5tools.py]:
```python
type(self.foo2.my_data.size)  # np.int64
self.assertEqual(np.int64(5), (5,))  # True
```

File: src/hdmf/backends/hdf5/h5_utils.py [hunk 5]
```diff
@@ -494,6 +503,9 @@ def __init__(self, **kwargs):
         call_docval_func(super().__init__, kwargs)
         # Construct the dict with the io args, ignoring all options that were set to None
         self.__iosettings = {k: v for k, v in zip(ioarg_names, ioarg_values) if v is not None}
+        if self.data is None:
+            self.__iosettings['dtype'] = self.dtype
+            self.__iosettings['shape'] = self.shape
```
Comment (oruebel) [inline src/hdmf/backends/hdf5/h5_utils.py:508]: Since ``dtype`` and ``shape`` are only allowed if ``data`` is ``None`` I would suggest adding a check here to raise a ``ValueError`` if the user tries to provide ``data`` and change the ``dtype`` or ``shape``.  We should then also add a unit test to check for those errors. 

```suggestion
        if self.data is None:
            self.__iosettings['dtype'] = self.dtype
            self.__iosettings['shape'] = self.shape
        else:
            if getargs('dtype', kwargs) is not None:
                 raise ValueError("Setting the dtype when data is not None is not supported")
            if  getargs('shape', kwargs) is not None:
                 raise ValueError("Setting the shape when data is not None is not supported")
```

**Review by oruebel — COMMENTED** [review]
Overall this looks good to me.

File: src/hdmf/backends/hdf5/h5_utils.py (hunk 5 above)
Comment (ajtritt) [reply src/hdmf/backends/hdf5/h5_utils.py:508]: @oruebel I moved this to DataIO. What do you think?

Comment (oruebel) [reply src/hdmf/backends/hdf5/h5_utils.py:508]: Sounds reasonable. In that case I would update the doc for those parameters in `DataIO` as well to mention this behavior.

File: src/hdmf/data_utils.py [hunk 6]
```diff
@@ -907,10 +907,19 @@ class DataIO:
     used to pass dataset-specific I/O parameters to the particular HDMFIO backend.
     """
 
-    @docval({'name': 'data', 'type': 'array_data', 'doc': 'the data to be written', 'default': None})
+    @docval({'name': 'data', 'type': 'array_data', 'doc': 'the data to be written', 'default': None},
+            {'name': 'dtype', 'type': (type, np.dtype), 'doc': 'the data type of the dataset', 'default': None},
+            {'name': 'shape', 'type': tuple, 'doc': 'the shape of the dataset', 'default': None})
```
Comment (oruebel) [inline src/hdmf/data_utils.py]:
```suggestion
    @docval({'name': 'data', 
             'type': 'array_data', 
             'doc': 'the data to be written', 
             'default': None},
            {'name': 'dtype', 
             'type': (type, np.dtype), 
             'doc': 'the data type of the dataset. Not used if data is specified.', 
             'default': None},
            {'name': 'shape', 
             'type': tuple, 
             'doc': 'the shape of the dataset. Not used if data is specified.', 
             'default': None})
```

**Review by oruebel — DISMISSED** [review]

**Review by oruebel — DISMISSED** [review]

File: tests/unit/test_io_hdf5_h5tools.py (hunk 4 above)
Comment (rly) [reply tests/unit/test_io_hdf5_h5tools.py]: This is probably why. A little gotcha when comparing numpy ints with other types
```python
>>> 5 == (5, )
False
>>> np.int64(5) == (5, )
array([ True])
>>> bool(np.int64(5) == (5,))
True
```

Comment (rly) [inline tests/unit/test_io_hdf5_h5tools.py]:
```suggestion
            self.assertEqual(self.foo2.my_data.shape, (5,))  # test getattr with shape
```
Can we revert this change?

File: tests/unit/test_io_hdf5_h5tools.py [hunk 7]
```diff
@@ -1595,8 +1595,8 @@ def test_link(self):
         with self.assertRaisesWith(InvalidDataIOError, "Cannot get length of data. Data is not valid."):
             len(self.foo2.my_data)
 
-        with self.assertRaisesWith(InvalidDataIOError, "Cannot get attribute 'shape' of data. Data is not valid."):
-            self.foo2.my_data.shape
+        with self.assertRaisesWith(InvalidDataIOError, "Cannot get attribute 'size' of data. Data is not valid."):
+            self.foo2.my_data.size
```
Comment (rly) [inline tests/unit/test_io_hdf5_h5tools.py]:
```suggestion
        with self.assertRaisesWith(InvalidDataIOError, "Cannot get attribute 'shape' of data. Data is not valid."):
            self.foo2.my_data.shape
```
And this change.

File: src/hdmf/backends/hdf5/h5tools.py [hunk 8]
```diff
@@ -1319,6 +1325,35 @@ def __setup_chunked_dset__(cls, parent, name, data, options=None):
...
+            raise ValueError(f"Cannot setup empty dataset {pp(parent.name, name)} without shape")
+        if 'dtype' not in io_settings:
+            raise ValueError(f"Cannot setup empty dataset {pp(parent.name, name)} without dtype")
+        if isinstance(io_settings['dtype'], str):
+            # map to real dtype if we were given a string
+            io_settings['dtype'] = cls.__dtypes.get(io_settings['dtype'])
+        try:
+            dset = parent.create_dataset(name, **io_settings)
+        except Exception as exc:
+            raise Exception("Could not create dataset %s in %s" % (name, parent.name)) from exc
```
Comment (rly) [inline src/hdmf/backends/hdf5/h5tools.py:1354]: Can you add a test for this exception, e.g., you try to create an empty dataset with a name, and the dataset already exists or dtype=object.

**Comment by rly** [comment]
Please fix flake8 changes and add a note to the changelog

**Review by rly — APPROVED** [review]

**Comment by rly** [comment]
Thanks. LGTM!
